plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java

```diff
@@ -101,6 +101,7 @@
 import static java.util.Objects.requireNonNull;
 
 public class CheckpointEntryIterator
```

Comment:
This iterator should be closeable because it opens a resource which might be left open if the iterator doesn't process all its entries.

Probably the code of the connector needs a bigger change to accustom this requirement.

Reply:
So this PR fixes the resource leak when CheckpointEntryIterator is exhausted, which is a good change & let's ship it.

The other question is how to ensure resource is closed when CheckpointEntryIterator is not exhausted (ie due to exception). Returning `Stream` is not sufficient to guarantee that, but can be helpful. From the class docs

> Streams have a [BaseStream.close()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/BaseStream.html#close()) method and implement [AutoCloseable](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/AutoCloseable.html). Operating on a stream after it has been closed will throw [IllegalStateException](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/IllegalStateException.html). Most stream instances do not actually need to be closed after use, as they are backed by collections, arrays, or generating functions, which require no special resource management. Generally, only streams whose source is an IO channel, such as those returned by [Files.lines(Path)](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#lines(java.nio.file.Path)), will require closing. If a stream does require closing, it must be opened as a resource within a try-with-resources statement or similar control structure to ensure that it is closed promptly after its operations have completed.

so yes, Stream + twr will be sufficient.

Reply:
Yes, I actually tried the approach first and changed to this AbstractIterator as it required more work. I will take another look when I have the time.